src/analysis/retail/warlock/destruction/modules/talents/Havoc.tsx

```diff
@@ -19,21 +22,48 @@ class Havoc extends Analyzer {
   protected enemies!: Enemies;
 
   damage = 0;
+  talent: Talent = TALENTS.HAVOC_TALENT;
+
+  HAVOCABLE_ABILITY_IDS = [
+    SPELLS.CHAOS_BOLT,
+    SPELLS.INCINERATE,
+    SPELLS.CONFLAGRATE,
+    TALENTS.SOUL_FIRE_TALENT,
+    TALENTS.SOULBURN_TALENT,
+    SPELLS.IMMOLATE,
+  ].map(s => s.id)
 
   constructor(options: Options) {
     super(options);
-    this.addEventListener(Events.damage.by(SELECTED_PLAYER), this.onDamage);
+    if (this.selectedCombatant.hasTalent(TALENTS.MAYHEM_TALENT)) {
+      this.talent = TALENTS.MAYHEM_TALENT;
+    }
+    this.active = this.selectedCombatant.hasTalent(TALENTS.HAVOC_TALENT) || this.selectedCombatant.hasTalent(TALENTS.MAYHEM_TALENT);
+
+    if (this.active) {
+      this.addEventListener(Events.damage.by(SELECTED_PLAYER), this.onDamage);
+    }
   }
 
   onDamage(event: DamageEvent) {
     const enemy = this.enemies.getEntity(event);
```

Comment:
If this function is only supposed to look at your havocable abilities, then consider using your HAVOCABLE_ABILITY_IDS const differently. Define it as a spell array that you can add to the event listener (essentially just remove the .map()): `this.addEventListener(Events.damage.by(SELECTED_PLAYER).spell([HAVOCABLE_ABILITIES]), this.onDamage);`
map your spellIds later on if you still need them

Havoc would probably be better handled using event linking so that you could easily filter out the hardcasted damage  vs the havoc damage to a target, probably out of scope for this PR though

Reply:
ah thanks for this tip, I didn't check if I could pass a list of spells for damage events. What can I look at to learn about event linking? I would like to revisit the wl spec modules after I do a pass on each one to make sure at least what already exists is updated for DF mechanics and add new things or tighten up calculations like this, so yeah I wouldn't edit it in this PR but I'd queue it up for a later one.

Reply:
There are some good examples in the Mistweaver Monk and Preservation Evoker files, filenames - CastLinkNormalizer.ts
Pres might be better to look at since most of the monk ones are used for hot tracking attribution.
My current PR for RSham https://github.com/WoWAnalyzer/WoWAnalyzer/pull/5872 also has a lot of event linking implementation in it as well

Reply:
Thanks, I will have a look. In the meantime I have swapped to passing the ability list into the listener as described.